Define an operator for OLM and SDK -- OSDK1656 (#115)
operator-framework/olm-docs: content/en/docs/operator.md

---
title: Operators
menu:
  main:
    weight: 10
---

## What is an Operator?

> An Operator is a method of packaging, deploying and managing a Kubernetes-native application. A Kubernetes-native application is an application that is both deployed on Kubernetes and managed using the Kubernetes APIs and kubectl tooling. An Operator is a piece of software that runs in a Pod inside the Kubernetes cluster and it interacts with the Kubernetes API. It has a reconciler loop that runs continuously, which monitors Objects (Pods, Services, ConfigMaps, or PersistentVolumes) for a desired state given by a Custom Resource Definition (CRD) and replicates that state in the cluster.
> 
> Conceptually, Operators take human operational knowledge and encode it into software that is more easily shared with consumers.
> 
> Operators are pieces of software that ease the operational complexity of running another piece of software. They act like an extension of the software vendor’s engineering team, watching over a Kubernetes environment (such as OpenShift Container Platform) and using its current state to make decisions in real time. Advanced Operators are designed to handle upgrades seamlessly, react to failures automatically, and not take shortcuts, like skipping a software backup process to save time.
> 
> More technically, Operators are a method of packaging, deploying, and managing a Kubernetes application.
> 
> A Kubernetes application is an app that is both deployed on Kubernetes and managed using the Kubernetes APIs and `kubectl` or `oc` tooling. To be able to make the most of Kubernetes, you require a set of cohesive APIs to extend in order to service and manage your apps that run on Kubernetes. Think of Operators as the runtime that manages this type of app on Kubernetes.

Reviewer: I am not sure if I agree with all that is defined here. And then, IMO we should leverage more in K8S docs instead of trying to create our own definitions in this case. Following my suggestion.

```suggestion
Operators make it easy to manage complex stateful applications on top of Kubernetes. By [Extendig the Kubernetes API with CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) it is possible to create what we call Operator Projects where you will be able to create and define your own API's to represent your solutions on the cluster and managed them by looking for the desired state. For a better understanding see [Operator Pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)
```

## What is the difference between a Controller and an Operator

> An Operator is an application-specific controller that extends the Kubernetes API to create, configure and manage instances of complex stateful applications on behalf of a Kubernetes user. It is built on top of the basic Kubernetes resource and Kubernetes Controller along with some application-specific knowledge.
> 
> We use Operators because managing stateful applications, like databases, caches and monitoring systems, is a big challenge, especially at massive scale. These systems require human operational knowledge to correctly scale, upgrade and reconfigure while at the same time protecting against data loss and unavailability.
> 
> All Operators use the controller pattern, but the converse is not true. It's only an Operator if it has API extension and single-app focus in addition to the controller pattern

Reviewer: Same here. IMO,  how much more we use the K8S docs and its projects docs to describe is better. Following my suggestion.

```suggestion
See that following the [Operator's pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/); you will create [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/) which provides a reconcile function responsible for synchronizing the resources until a desired state on the cluster. For a practical explanation see also [What’s in a controller?](https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller).
```

> ## What do Operators provide?
> 
> - Repeatability of installation and upgrade.
> - Constant health checks of every system component.
> - Over-the-air (OTA) updates for Kubernetes components and ISV content.
> - A place to encapsulate knowledge from field engineers and spread it to all users, not just one or two.

Reviewer: IMHO we do not need to add this one. The link for a further explanation was added above https://kubernetes.io/docs/concepts/extend-kubernetes/operator.

```suggestion
```

## What is Operator Lifecycle Manager?

> Operator Lifecycle Manager (OLM) helps users install, update, and manage the lifecycle of all Operators and their associated services running across their clusters. It is part of the Operator Framework, an open source toolkit designed to manage Kubernetes native applications (Operators) in an effective, automated, and scalable way.
> 

Reviewer: ```suggestion
Operator Lifecycle Manager (OLM) helps users install, update, and manage the lifecycle of all Operators and their associated services running across their clusters. This project is a component of the [Operator Framework](https://github.com/operator-framework), an open-source toolkit to manage Kubernetes native applications, called Operators, in an effective, automated, and scalable way. Read more in the introduction [blog post](https://www.openshift.com/blog/introducing-the-operator-framework?extIdCarryOver=true&sc_cid=701f2000001Css5AAC).
```

## How does OLM enable cluster admins and Developers?

The OLM is enabled by default in OpenShift Container Platform 4.X, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.

For developers, a self-service experience allows provisioning and configuring instances of databases, monitoring, and big data services without having to be subject matter experts, because the Operator has that knowledge baked into it.

## Build with the Operator SDK

> The [Operator SDK](https://sdk.operatorframework.io/) provides the tools to build, test and package Operators. The Operator SDK strips away a lot of the boilerplate code that is normally required to integrate with the Kubernetes API. It also provides a usable scaffolding so developers can focus on adding business logic (for example, how to scale, upgrade, or backup the application it manages). Leading practices and code patterns shared across Operators are included in the Operator SDK to help prevent duplicating efforts. The Operator SDK also encourages short, iterative development and test cycles with tooling that allow for basic validation of the Operator, and automated packaging for deployment using the [Operator Lifecycle Manager](https://olm.operatorframework.io/).

Reviewer: ```suggestion
[Operator SDK](https://sdk.operatorframework.io/) is also a component of the [Operator Framework](https://github.com/operator-framework). Writing an operator today can be difficult because of challenges such as using low-level APIs, writing boilerplate, and a lack of modularity which leads to duplication.

The [Operator SDK](https://sdk.operatorframework.io/) is a framework that uses the [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) library to make writing operators easier by providing:

- High-level APIs and abstractions to write the operational logic more intuitively
- Tools for scaffolding and code generation to bootstrap a new project fast
- Extensions to cover common operator use cases

Note that [Operator SDK](https://sdk.operatorframework.io/) provides helpers and features to help you [integrate your project with OLM](https://sdk.operatorframework.io/docs/olm-integration/).  For further information check https://sdk.operatorframework.io/ .

```

## Package with the Operator Lifecycle Manager

With OLM, administrators can control which Operators are available in what namespaces and who can interact with running Operators. The permissions of an Operator are accurately configured automatically to follow a least-privilege approach. OLM manages the overall lifecycle of Operators and their resources, by doing things like resolving dependencies on other Operators, triggering updates to both an Operator and the application it manages, or granting a team access to an Operator for their slice of the cluster.

Simple, stateless applications can use the Lifecycle Management features of the Operator Framework—without writing any code—by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating such features as updates, backups and scaling.

> 

Reviewer: instead of using the links as `[Operator Framework](https://github.com/operator-framework)` the standard for docs is to create an alias and add the ref on the bottom of the page. e.g see the end oof the file; https://raw.githubusercontent.com/operator-framework/operator-sdk/master/website/content/en/docs/building-operators/golang/migration.md 

You will replace `[Operator SDK](https://sdk.operatorframework.io/)` with `[Operator SDK][operator-sdk]` and add at the end of the file `[operator-sdk]: https://sdk.operatorframework.io/`

Author: I have replaced all repetitive links in this format

#### Next steps

> 

Reviewer: Also, for the specific OLM questions, I think would be nice to get a review from an OLM team member.

Author: Thanks for your feedback!

> #### Next steps
> 
> See [How to use Operator SDK to build operators](https://sdk.operatorframework.io/build/). And then, check how Operator SDK can help you to [Integrate your Operator with OLM](https://sdk.operatorframework.io/docs/olm-integration/)

Reviewer: if you follow/accepted my above suggestions it shows no longer required. 
```suggestion
```

Conceptually, Operators take human operational knowledge and encode it into software that is more easily shared with consumers.

Operators are pieces of software that ease the operational complexity of running another piece of software. They act like an extension of the software vendor’s engineering team, watching over a Kubernetes environment (such as OpenShift Container Platform) and using its current state to make decisions in real time. Advanced Operators are designed to handle upgrades seamlessly, react to failures automatically, and not take shortcuts, like skipping a software backup process to save time.

More technically, Operators are a method of packaging, deploying, and managing a Kubernetes application.

A Kubernetes application is an app that is both deployed on Kubernetes and managed using the Kubernetes APIs and `kubectl` or `oc` tooling. To be able to make the most of Kubernetes, you require a set of cohesive APIs to extend in order to service and manage your apps that run on Kubernetes. Think of Operators as the runtime that manages this type of app on Kubernetes.

## What is the difference between a Controller and an Operator

> An Operator is an application-specific controller that extends the Kubernetes API to create, configure and manage instances of complex stateful applications on behalf of a Kubernetes user. It is built on to of the basic Kubernetes resource and Kubernetes Controller along with some application-specific knowledge.

Author: Applied the changes

We use Operators because managing stateful applications, like databases, caches and monitoring systems, is a big challenge, especially at massive scale. These systems require human operational knowledge to correctly scale, upgrade and reconfigure while at the same time protecting against data loss and unavailability.

All Operators use the controller pattern, but the converse is not true. It's only an Operator if it has API extension and single-app focus in addition to the controller pattern

> 

Reviewer: Is this content from any other place and we are copying and past here?
I think we have similar content in SDK docs as well. Then, I am wondering if would not be possible we ensure that both places are with the same content.

Author: I have got this content from an external link. Could you please tell me in which place is it available in our docs? I will make sure the content is exactly the same

Reviewer: here: https://sdk.operatorframework.io/docs/overview/

Reviewer: also, see here the explanation about what is a controller https://book.kubebuilder.io/cronjob-tutorial/controller-overview.html#whats-in-a-controller.

## What do Operators provide?

- Repeatability of installation and upgrade.
- Constant health checks of every system component.
- Over-the-air (OTA) updates for OpenShift components and ISV content.

> - A place to encapsulate knowledge from field engineers and spread it to all users, not just one or two.

Reviewer: We cannot use `OpenShift` in the public docs. We have been trying to avoid that.

Author: I have changed the Openshit to Kubernetes

## What is Operator Lifecycle Manager?

Operator Lifecycle Manager (OLM) helps users install, update, and manage the lifecycle of all Operators and their associated services running across their clusters. It is part of the Operator Framework, an open source toolkit designed to manage Kubernetes native applications (Operators) in an effective, automated, and scalable way.

## How does OLM enable cluster admins and Developers?

The OLM is enabled by default in OpenShift Container Platform 4.X, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.

For developers, a self-service experience allows provisioning and configuring instances of databases, monitoring, and big data services without having to be subject matter experts, because the Operator has that knowledge baked into it.

## Build with the Operator SDK

> The [Operator Software Development Kit (SDK)](https://sdk.operatorframework.io/) provides the tools to build, test and package Operators. The SDK strips away a lot of the boilerplate code that is normally required to integrate with the Kubernetes API. It also provides a useable scaffolding so developers can focus on adding business logic (for example, how to scale, upgrade, or backup the application it manages). Leading practices and code patterns shared across Operators are included in the SDK to help prevent duplicating efforts. The SDK also encourages short, iterative development and test cycles with tooling that allow for basic validation of the Operator, and automated packaging for deployment using the [Operator Lifecycle Manager](https://olm.operatorframework.io/).

Author: Applied the changes

## Package with the Operator Lifecycle Manager

> The Operator Lifecycle Manager (OLM) is the backplane that facilitates management of operators on a Kubernetes cluster. Operators that provide popular applications as a service are going to be long-lived workloads with, potentially, lots of permissions on the cluster.

Author: I have removed the redundant portion

With OLM, administrators can control which Operators are available in what namespaces and who can interact with running Operators. The permissions of an Operator are accurately configured automatically to follow a least-privilege approach. OLM manages the overall lifecycle of Operators and their resources, by doing things like resolving dependencies on other Operators, triggering updates to both an Operator and the application it manages, or granting a team access to an Operator for their slice of the cluster.

Simple, stateless applications can use the Lifecycle Management features of the Operator Framework—without writing any code—by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating such features as updates, backups and scaling.

#### Next steps

> See [How to use SDK to build operators](https://sdk.operatorframework.io/build/). And then, check how SDK can help you to [Integrate your Operator with OLM](https://sdk.operatorframework.io/docs/olm-integration/)

Author: Applied the changes!

All Operators use the controller pattern, but the converse is not true. It's only an Operator if it has API extension and single-app focus in addition to the controller pattern

## What do Operators provide?

- Repeatability of installation and upgrade.
- Constant health checks of every system component.
- Over-the-air (OTA) updates for OpenShift components and ISV content.
- A place to encapsulate knowledge from field engineers and spread it to all users, not just one or two.

## What is Operator Lifecycle Manager?

Operator Lifecycle Manager (OLM) helps users install, update, and manage the lifecycle of all Operators and their associated services running across their clusters. It is part of the Operator Framework, an open source toolkit designed to manage Kubernetes native applications (Operators) in an effective, automated, and scalable way.

## How does OLM enable cluster admins and Developers?

> The OLM runs by default in OpenShift Container Platform 4.1, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.

Reviewer: ```suggestion
The OLM is enabled by default in any OpenShift Container Platform 4.X, which aids cluster administrators in installing, upgrading, and granting access to Operators running on their cluster. The OpenShift Container Platform web console provides management screens for cluster administrators to install Operators, as well as grant specific projects access to use the catalog of Operators available on the cluster.
```

Author: I have made the changes

For developers, a self-service experience allows provisioning and configuring instances of databases, monitoring, and big data services without having to be subject matter experts, because the Operator has that knowledge baked into it.

## Build with the Operator SDK

The Operator software development kit (SDK) provides the tools to build, test and package Operators. The SDK strips away a lot of the boilerplate code that is normally required to integrate with the Kubernetes API. It also provides a useable scaffolding so developers can focus on adding business logic (for example, how to scale, upgrade, or backup the application it manages). Leading practices and code patterns shared across Operators are included in the SDK to help prevent duplicating efforts. The SDK also encourages short, iterative development and test cycles with tooling that allow for basic validation of the Operator, and automated packaging for deployment using the Operator Lifecycle Manager.

> 

Reviewer: would be nice add here the links for the SDK website as well.

Author: I have added some links

With OLM, administrators can control which Operators are available in what namespaces and who can interact with running Operators. The permissions of an Operator are accurately configured automatically to follow a least-privilege approach. OLM manages the overall lifecycle of Operators and their resources, by doing things like resolving dependencies on other Operators, triggering updates to both an Operator and the application it manages, or granting a team access to an Operator for their slice of the cluster.

Simple, stateless applications can use the Lifecycle Management features of the Operator Framework—without writing any code—by using a generic Operator (for example, the Helm Operator). However, complex and stateful applications are where an Operator can be especially useful. The managed-service capabilities that are encoded into the Operator code can provide an advanced user experience, automating such features as updates, backups and scaling.

> #### Follow the links to set up sample Operators and integrate it with OLM

Reviewer: ```suggestion
#### Next steps
```

> [Create sample operator](https://sdk.operatorframework.io/docs/building-operators/)

Reviewer: ```suggestion
- See [How to use SDK to build operators](https://sdk.operatorframework.io/build/)
```

Author: Changes have been applied

> [Integrate your Operator with OLM](https://sdk.operatorframework.io/docs/olm-integration/)

Reviewer: ```suggestion
- And then, check how SDK can help you to [Integrate your Operator with OLM](https://sdk.operatorframework.io/docs/olm-integration/)
```
